node/cn/handler_test.go

```diff
@@ -58,3 +62,169 @@ func TestNewProtocolManager(t *testing.T) {
+	hash := block.Hash()
+
+	// 1. When the given node is a consensus node.
+	{
+		// len(pnsWithoutBlock) <= blockReceivingPNLimit
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.CONSENSUSNODE
+		pm.peers = mockPeers
+
+		mockCN := NewMockPeer(mockCtrl)
+		mockPN := NewMockPeer(mockCtrl)
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return([]Peer{mockPN}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 2, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
+		assert.Equal(t, mockPN, sampledPeers[1])
+
+		// len(pnsWithoutBlock) > blockReceivingPNLimit
+		mockPeers = NewMockPeerSet(mockCtrl)
+		pm.peers = mockPeers
+
+		mockCN = NewMockPeer(mockCtrl)
+		mockPN = NewMockPeer(mockCtrl)
+
+		var pnWithoutBlock []Peer
+		for i := 0; i <= blockReceivingPNLimit; i++ {
+			pnWithoutBlock = append(pnWithoutBlock, mockPN)
+		}
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return(pnWithoutBlock).Times(1)
+
+		sampledPeers = pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1+blockReceivingPNLimit, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
+	}
+
+	// 2. When the given node is a proxy node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.PROXYNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().PeersWithoutBlockExceptCN(hash).Return([]Peer{mockPeer}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1, len(sampledPeers))
+		assert.Equal(t, mockPeer, sampledPeers[0])
+	}
+
+	// 3. When the given node is an end point node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.ENDPOINTNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().ENWithoutBlock(hash).Return([]Peer{mockPeer}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1, len(sampledPeers))
+		assert.Equal(t, mockPeer, sampledPeers[0])
+	}
+
+	// 4. When the given node is a boot node.
+	{
+		pm.nodetype = node.BOOTNODE
+		assert.Equal(t, []Peer{}, pm.samplePeersToSendBlock(block))
+	}
+
+	// 5. When the given node is an unknown node.
+	{
+		pm.nodetype = node.UNKNOWNNODE
+		assert.Equal(t, []Peer{}, pm.samplePeersToSendBlock(block))
+	}
+}
+
+func TestSamplingPeers(t *testing.T) {
+	peers := make([]Peer, 10)
+	assert.Equal(t, peers, samplingPeers(peers, 20))
+	assert.Equal(t, peers[:5], samplingPeers(peers, 5))
+}
+
+func TestBroadcastBlock_NoParentExists(t *testing.T) {
+	pm := &ProtocolManager{}
+	pm.nodetype = node.ENDPOINTNODE
+	block := newBlock(123)
+	mockCtrl := gomock.NewController(t)
+
+	mockBlockChain := mocks.NewMockBlockChain(mockCtrl)
+	mockBlockChain.EXPECT().GetBlock(block.ParentHash(), block.NumberU64()-1).Return(nil).Times(1)
+
+	pm.blockchain = mockBlockChain
+	pm.BroadcastBlock(block)
```

Comment:
Don't you need to add assert statements to check the behavior?

Reply:
Why don't we add return error of `BroadcastBlock` and check the result for `Propagating dangling block` error?

Reply:
@KimKyungup This is because we don't handle the error outside `BroadcastBlock`. If an error happens it is handled inside the method and not propagated. As we don't have any handling logic at the caller of `BroadcastBlock`, we need to redesign the callers of `BroadcastBlock` if it returns any error.

Reply:
Then, How about checking if `AsyncSendNewBlockHash` call time is zero? if possible?

Reply:
Yes, I will update the test.